vernac/comCheck.ml

```diff
@@ -0,0 +1,90 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+type check_result = {
+  env : Environ.env;
+  sigma : Evd.evar_map;
+  judgment : EConstr.unsafe_judgment;
+  evars : Evar.Set.t;
+}
+
+type eval_result = {
+  env : Environ.env;
+  sigma : Evd.evar_map;
+  judgment : EConstr.unsafe_judgment;
+  redfun : Reductionops.reduction_function;
+  rc : Constrexpr.constr_expr;
+}
+
+type global_check_result = {
+  env : Environ.env;
+  sigma : Evd.evar_map;
```

Comment:
really weird that this has an evar map

Reply:
That's for the printing, the result may be open so you need to interpret evars in printing IIUC.

I was thinking of a simpler result type, but now that I see what Maxime has done it looks better this way to me. Ideally we would have something like a "printable object" record.

Reply:
>That's for the printing, the result may be open so you need to interpret evars in printing IIUC.

The whole point of global_check is that the result is not open.

Reply:
The code may be simplified if we have some invariants about terms being evar-free. Note that some operations on universes are also performed in the evar map.

Anyway, IMHO such improvements are orthogonal to this PR, which only exposes the current API (hopefully in a way that is explicit and robust enough). I just did a tiny bit of cleanup to remove the dummy half-factorization of `Eval` and `Check` which was very strange style and was making the types much less precise.

Reply:
>the dummy half-factorization of Eval and Check which was very strange style and was making the types much less precise.

Shouldn't it have been factorized more instead? AFAICT we return the reduction function and the judgment to call print_eval on them, but why don't we instead return the reduced judgment?

To the user the difference between `Eval red in term` and `Check ltac:(let c := open_constr:(term) in let c := eval red in c in refine c)` is approximately that Check prints info about evars and Eval doesn't, and maybe some subtleties about ltac that don't really matter since I only used ltac because there's no other way to Eval a subterm.

In other words `Check term` should be equivalent to `Eval identity in term`.

So IMO we should factorize more (and remove the differences we don't care about) instead of less.

>Anyway, IMHO such improvements are orthogonal to this PR, which only exposes the current API (hopefully in a way that is explicit and robust enough).

This is untrue, this PR clearly modifies the API. If you just want to expose the API then do that and splitting printing and compute can be done in a followup PR.

Reply:
Ok, I don't care about that part so I'll open a more uncontroversial PR.